src/platform/EFR32/Efr32PsaOperationalKeystore.cpp

```diff
@@ -0,0 +1,446 @@
+    if (error != CHIP_NO_ERROR)
+    {
+        return error;
+    }
+
+    // Initialize cache key
+    mCachedKey = Platform::New<EFR32OpaqueP256Keypair>();
+    if (!mCachedKey)
+    {
+        return CHIP_ERROR_NO_MEMORY;
+    }
+
+    mIsInitialized = true;
+    return CHIP_NO_ERROR;
+}
+
+bool Efr32PsaOperationalKeystore::HasOpKeypairForFabric(FabricIndex fabricIndex) const
+{
+    VerifyOrReturnError(mIsInitialized, false);
+    VerifyOrReturnError(IsValidFabricIndex(fabricIndex), false);
+
+    // If there was a pending keypair, then there's really a usable key
+    if (mIsPendingKeypairActive && (fabricIndex == mPendingFabricIndex) && (mPendingKeypair != nullptr))
+    {
+        return true;
+    }
+
+    // Scroll the map linearly
+    for (size_t i = 0; i < SL_MATTER_MAX_STORED_OP_KEYS; i++)
+    {
+        if (fabricIndex == mKeyMapping[i])
+        {
+            return true;
+        }
+    }
+
+    return false;
+}
+
+CHIP_ERROR Efr32PsaOperationalKeystore::NewOpKeypairForFabric(FabricIndex fabricIndex,
+                                                              MutableByteSpan & outCertificateSigningRequest)
+{
+    CHIP_ERROR error = CHIP_NO_ERROR;
+    VerifyOrReturnError(mIsInitialized, CHIP_ERROR_INCORRECT_STATE);
+    VerifyOrReturnError(IsValidFabricIndex(fabricIndex), CHIP_ERROR_INVALID_FABRIC_INDEX);
+
+    // If a key is pending, we cannot generate for a different fabric index until we commit or revert.
+    if ((mPendingFabricIndex != kUndefinedFabricIndex) && (fabricIndex != mPendingFabricIndex))
+    {
+        return CHIP_ERROR_INVALID_FABRIC_INDEX;
+    }
+
+    VerifyOrReturnError(outCertificateSigningRequest.size() >= Crypto::kMAX_CSR_Length, CHIP_ERROR_BUFFER_TOO_SMALL);
+
+    // Generate new key
+    EFR32OpaqueKeyId id = kEFR32OpaqueKeyIdUnknown;
+
+    if (mPendingFabricIndex != kUndefinedFabricIndex)
+    {
+        // If we already have a pending key, delete it and put a new one in its place
+        id = mPendingKeypair->GetKeyId();
+        if (id == kEFR32OpaqueKeyIdUnknown)
+        {
+            ResetPendingKey();
+        }
+        else
+        {
+            mPendingKeypair->Delete();
+            if (id == kEFR32OpaqueKeyIdVolatile)
+            {
+                id = kEFR32OpaqueKeyIdUnknown;
+            }
+        }
+    }
+
+    if (id == kEFR32OpaqueKeyIdUnknown)
+    {
+        // Find empty slot in keymap
+        for (size_t i = 0; i < SL_MATTER_MAX_STORED_OP_KEYS; i++)
+        {
+            if (mKeyMapping[i] == kUndefinedFabricIndex)
+            {
+                id = i + kEFR32OpaqueKeyIdPersistentMin;
+                break;
+            }
+        }
+
+        if (!mPendingKeypair)
+        {
+            mPendingKeypair = Platform::New<EFR32OpaqueP256Keypair>();
+        }
+    }
+
+    if (id < kEFR32OpaqueKeyIdPersistentMin || id > kEFR32OpaqueKeyIdPersistentMax)
```

Comment:
Should it be `>= kEFR32OpaqueKeyIdPersistentMax` ? Hard to say without more context

Reply:
Made the condition clearer by explicitly using the undefined key ID as marker